### ISSUE TITLE
Feature request: provide a back-to-original-page button

### ISSUE THREAD

**Issue by Chris75forumname** [NetaliDev/dokuwiki-deepl-autotranslate#2]
I would suggest something like the discussion button in the bootstrap 3 template, e.g. here: https://pepinature.org/fr/index
If you press the discussion button, it switches to the discussion page in the discussion namespace (just like a language namespace). If the page does not exist yet, it is created on the fly. A back-to-the-original-page button is provided.

**Comment by NetaliDev** [comment]
Sounds useful and isn't difficult to implement. I have added it to my TODO-List :)

**Comment by Chris75forumname** [comment]
Oh, sounds great! I was just about to do some hacking myself. I found the page_exists($ID) method. We would just check if the ID=en:namespace:page is populated and if yes, open that page through the link in the page menu. On the en-page the link is changed again to the original page. If the page does not exist, it is push created and opened (not in edit mode if option direct mode is toggled). 
But, I would appreciate very much to just be able to update the plugin because my programming skills are only beginner level :-)

**Comment by NetaliDev** [comment]
No problem :) I use the page_exists method on many points in the plugin and you had about the same idea as me.

But I think it is only possible to implement a back-button for going to the original language page since the plugin can translate pages into multiple languages and if you click the button on the original language page, you cannot decide to which translated page you should go.

But that shouldn't be a problem because you still have the language picker from the translate-plugin in your wiki (which is for example a part of the Bootstrap3 theme that you use).

**Comment by Chris75forumname** [comment]
I use the direct mode with push translation button and have only one other language code in the settings [en]. There was no picker shown to me so far, so it should be fine. Sorry for the confusion, the example page was from my friend. I host a intranet wiki, which can't be accessed online. I use the Sprintdoc Template. Just so you know. :-)
BTW: What would be even better, if the translated page is older than the original, the Push button could automatically translate and overwrite the old translation. You see, I am a fan of the direct mode! Cheers!

**Comment by NetaliDev** [comment]
> BTW: What would be even better, if the translated page is older than the original, the Push button could automatically translate and overwrite the old translation. You see, I am a fan of the direct mode! Cheers!

The push (or retranslate) button does exactly what you described right now (except that it does not check which page is older, it just overrides the translation(s)).

> I use the direct mode with push translation button and have only one other language code in the settings [en]. There was no picker shown to me so far, so it should be fine.

The picker has to be added to the template manually when it is not shipped with the template (like in the Bootstrap3 template as mentioned above). If you want to add the picker to your template, just follow the instructions from the "Installation"-Section of the [Translation-Plugin](https://www.dokuwiki.org/plugin:translation).

And sure it is possible to switch between the two languages if there only is the original language and one other language, but that's not the general use-case, because my DeepL-Plugin can translate into multiple languages and I'm not really happy with the idea of adding a button which only works if you have only one other language.

TLDR: A back-button would be fine and very easy to implement and I don't really like the idea of a "forward-button" which only works if you have only one other language.

But do you really need this when you have a working language-picker in your template?

**Comment by Chris75forumname** [comment]
Hi Jennifer,

> The push (or retranslate) button does exactly what you described right now (except that it does not check which page is older, it just overrides the translation(s)).

Okay is that also what the direct mode does? It would be good to first check if the original page has a new version since the last translation was triggered. So, we save some of the limited free characters of the DeepL API key.

> The picker has to be added to the template manually when it is not shipped with the template (like in the Bootstrap3 template as mentioned above). If you want to add the picker to your template, just follow the instructions from the "Installation"-Section of the [Translation-Plugin](https://www.dokuwiki.org/plugin:translation).

Yes, I am afraid I missed that part. This is why I didn't see a picker. I checked out the picker from the DokuWiki Homepage-Wiki. But, I prefer a button in the page menu.

> TLDR: A back-button would be fine and very easy to implement and I don't really like the idea of a "forward-button" which only works if you have only one other language.

Okay, I can easily do the "forward-button" myself by adding a button in the page menu and a click event to go to the translated page.

> But do you really need this when you have a working language-picker in your template?

Sorry, I think I got that wrong, since I did not test the picker yet, I did not know, it always translates on demand.

To modify my initial request, actually, it would be great if you could just add a check to execute the auto-re-translation only if the original page has a newer version than the translated page (in direct mode).

Thank you very much for your efforts and your care!

Cheers!

**Comment by NetaliDev** [comment]
> Hi Jennifer,
> 
> > The push (or retranslate) button does exactly what you described right now (except that it does not check which page is older, it just overrides the translation(s)).
> 
> Okay is that also what the direct mode does? It would be good to first check if the original page has a new version since the last translation was triggered. So, we save some of the limited free characters of the DeepL API key.

The direct mode only translates it the page in the other language does not exist. So if you re-visit the page, it won't be translated again until you delete it or press the retranslate button.

> 
> > The picker has to be added to the template manually when it is not shipped with the template (like in the Bootstrap3 template as mentioned above). If you want to add the picker to your template, just follow the instructions from the "Installation"-Section of the [Translation-Plugin](https://www.dokuwiki.org/plugin:translation).
> 
> Yes, I am afraid I missed that part. This is why I didn't see a picker. I checked out the picker from the DokuWiki Homepage-Wiki. But, I prefer a button in the page menu.

It would be possible to add a forward-button when I add a plugin config option for the target language of the forward button. And as I've already said a back-button is easy in any case.

> > TLDR: A back-button would be fine and very easy to implement and I don't really like the idea of a "forward-button" which only works if you have only one other language.
> 
> Okay, I can easily do the "forward-button" myself by adding a button in the page menu and a click event to go to the translated page.

As described above, I can add a forward-button with a config option.

> > But do you really need this when you have a working language-picker in your template?
> 
> Sorry, I think I got that wrong, since I did not test the picker yet, I did not know, it always translates on demand.
> 
> To modify my initial request, actually, it would be great if you could just add a check to execute the auto-re-translation only if the original page has a newer version than the translated page (in direct mode).

As already said above, the direct-mode does nothing to existing pages. You can only retranslate them manually by pressing the retranslate button or by deleting the page. But you have a good idea with the on-demand retranslations when the original page has changed, I will add that as a config option!

> Thank you very much for your efforts and your care!

And thank you for your input :)

**Comment by Chris75forumname** [comment]
Hi Jennifer,
I have one more idea how to easily implement these on-demand retranslations. If the config option is set accordingly, all that has to be done is, whenever a page is edited and saved, the plugin has to check if an associated language page exists and delete it (because it is now expired / not up-to-date anymore). Everything else does not need to be touched. Sounds simple, doesn't it??

**EDIT:** 
- https://www.dokuwiki.org/devel:event:common_wikipage_save (relevant event)
- https://codesearch.dokuwiki.org/xref/plugin/deletepagebutton/action.php#95 (example code to delete / save empty page)
- https://codesearch.dokuwiki.org/xref/plugin/log/common.php#80 (example code to save content to another page)

**Comment by NetaliDev** [comment]
Uhh yes that's a great idea!

**Comment by Chris75forumname** [comment]
> But do you really need this when you have a working language-picker in your template?

Hi, I have come to appriciate the Language Code Buttons from the Translation Plugin in their simplest form, just language code (without flag or title). They appear in the page header in the left hand corner, which is fine. So I don't need the buttons in the page menu, anymore (push and back buttons). 

> But you have a good idea with the on-demand retranslations when the original page has changed, I will add that as a config option!

Yes, it would still be great to have the option to have language files deleted with page edits. So the direct mode translations are always up-to-date. `Since this is a separate topic I open an new thread`
